### ISSUE TITLE
HalfTensor to/from numpy

### ISSUE THREAD

**Issue by narendasan** [pytorch/pytorch#2014]
Is it possible to convert a HalfTensor to/from a numpy array like you can with FloatTensor? If not I would be happy to work on adding it

Also a separate point the documentation seems to be out of date on the existence of HalfTensor on the cpu

**Comment by soumith** [comment]
> Is it possible to convert a HalfTensor to/from a numpy array like you can with FloatTensor? If not I would be happy to work on adding it

No. Does NumPy have a half type that's exactly like a CUDA half type?

> Also a separate point the documentation seems to be out of date on the existence of HalfTensor on the cpu

We'd love a fix for this (thank you! :) )

**Comment by narendasan** [comment]
From my understanding numpy.half is compliant with the IEEE 754-2008 binary16 spec. I believe that CUDA's half follows the same standard so it should work in theory (will try and verify this for sure). 

Cursory google search:
https://devblogs.nvidia.com/parallelforall/new-features-cuda-7-5/
https://github.com/numpy/numpy/blob/7cfec2403486456b52b525eccf7541e1562d9ab3/doc/release/1.6.0-notes.rst#new-16-bit-floating-point-type

Will submit a patch for the documentation

**Comment by soumith** [comment]
if numpy half and cuda half are compatible, then sure, we can make torch.from_numpy work with numpy half. The code change will be here:
https://github.com/pytorch/pytorch/blob/d9d50f80c7f162c9c6c6d95c0b51f9e6fd8fccee/torch/csrc/Module.cpp#L154 `for from_numpy`
https://github.com/pytorch/pytorch/blob/08648061f7690f6fad817677006b7159978db08d/torch/csrc/generic/Tensor.cpp#L7 `for .numpy()`

Thank you

**Comment by narendasan** [comment]
I have ```from_numpy``` working but ```.numpy()``` doesn't seem to work by just adding 
```
#ifdef TH_REAL_IS_HALF
#define NUMPY_TYPE_ENUM NPY_HALF
#endif
```

I was looking around and there seems some specific HalfTensor code, could that be effecting the code gen?

I also found that the RuntimeError you get when using an incorrect type in trying to make a HalfTensor refers to the tensor as an int tensor instead of a float tensor 
```
RuntimeError: tried to construct a tensor from a int sequence, but found an item of 
type numpy.float32 at index (0)
```
I saw where the ```python_type_str``` is defined in ```utils.h``` it refers to ```THC_REAL_IS_HALF``` instead of ```TH_REAL_IS_HALF```. What is the difference between the two symbols and should I be using one over the other? 

Also I noticed that there are some inconsistencies between DoubleTensor, FloatTensor and HalfTensor.
Creating a Tensor with an FP16/FP32 numpy array only works for half and float respectively but FP64 works for all of them, is this expected?

**Comment by soumith** [comment]
if you push your changes to a PR, I can have a look.

**Comment by narendasan** [comment]
Will create a PR tonight.

For the FP64 issue, when you try to create a ```HalfTensor``` with an FP32 array, you get an error which is expected, when you try to create a ```FloatTensor``` with a FP16 array, you get and error as well, and if you try to provide a FP16 or FP32 array to a ```DoubleTensor```, both fail. But for both ```HalfTensor``` and ```FloatTensor```  FP64 arrays are accepted. Is this expected? 

example:
```
>>> a = np.ones(3, np.float64)
>>> torch.DoubleTensor(a)

 1
 1
 1
[torch.DoubleTensor of size 3]

>>> torch.HalfTensor(a)

 1
 1
 1
[torch.HalfTensor of size 3]

>>> torch.FloatTensor(a)

 1
 1
 1
[torch.FloatTensor of size 3]
```

**Comment by apaszke** [comment]
No, it's a bug.

**Comment by narendasan** [comment]
Should a fix for that be included in the patch for this or should it be in a separate PR?

**Comment by apaszke** [comment]
Both ways are fine 🙂

**Comment by narendasan** [comment]
added this pr for the Numpy support, https://github.com/pytorch/pytorch/pull/2123 will create a new one for the bug

**Comment by narendasan** [comment]
As an update to this issue, I got pretty busy with work for the last month, I am off now so I am going to be picking this back up if it hasn't been addressed yet

**Comment by niffler92** [comment]
Hi, has this issue been fixed in the future releases?

**Comment by vishwakftw** [comment]
I think this can be closed due to #2953 .

**Comment by soumith** [comment]
closed via #2953
